### ISSUE TITLE
xl: multipart timesout with 'mc' with 16GiB dataset

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#1350]
The problem here is that we are spending way too much time concatenating and writing to disk.

```
minio@minio1:~/gopath/src/github.com/minio/mc$ mc session resume CAuAiUOc
mc: <ERROR> Failed to copy ‘/home/minio/data-sets/ol_cdump_latest.txt.gz’. Post http://localhost:9000/testbucket/ol_cdump_latest.txt.gz?uploadId=ce657481-2d54-47ef-9b7b-b2d1adde5812: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
mc: <ERROR> Session safely terminated. To resume session ‘mc session resume CAuAiUOc’
minio@minio1:~/gopath/src/github.com/minio/mc$
```

This happens even with 

```
$ git grep Timeout
client-s3.go:           api.SetClientTimeout(8 * time.Minute)
```

We need to see how to fix these issues.

**Comment by harshavardhana** [comment]
Assigned to @krishnasrinivas

**Comment by harshavardhana** [comment]
Should be fixed by #1179
